### PR TITLE
Moving fps and cur to scikit-cosmo version with backwards compatibility

### PR DESCRIPTION
**Pull request by max-veit**
The FPSFilter and CURFilter classes currently implement functionality that is better handled (in some cases, already implemented) in scikit-cosmo. Therefore, we are refactoring these classes so that these tasks can be handled by skcosmo instead.

This refactoring includes removing some code duplication that was present in these versions (and caused e.g. #333 ) by making both filters inherit from a common base class, `Filter`. The new `CURFilter` and `FPSFilter` classes behave exactly as the old ones, so there should be little to no change necessary for end users.

Tasks before review:

- [ ] Add tests, examples, and complete documentation of any added functionality
    - [ ] Make sure the autogenerated documentation appears in Sphinx and is
          formatted correctly (ask @max-veit if you need help with this task).
    - [ ] Write additional documentation in the Sphinx (not docstrings!) to
          explain the feature and its usage in plain English
    - [ ] Make sure the examples run (!) and produce the expected output
- [ ] Run `make lint` on the project, ensure it passes
    - [ ] Review variable names, make sure they're descriptive (ask a friend)
    - [ ] Review all TODOs, convert to issues wherever possible
    - [ ] Make sure draft, in-progress, and commented-out code is moved to its
          own branch
- [ ] If committing any code in Jupyter/IPython notebooks, install and run `nbstripout`
- [ ] Merge with master and resolve any conflicts
- [ ] Make scikit-cosmo dependency optional?

### CODE REVIEW

**Comment by Luthaf** [comment]
This PR looks very goo overall, the new code is much simpler!

File: bindings/rascal/utils/__init__.py [hunk 1]
```diff
@@ -10,12 +10,11 @@
 
 # Warning potential dependency loop: FPS imports models, which imports KRR,
 # which imports this file again
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/__init__.py:12]: Ah yes, I didn't even realize that we'd basically solved #294 a different way by eliminating our homebrew FPS solution!

Speaking of which... maybe we should get rid of that code (FPS module and C++ implementation of FPS), seeing as it's no longer used.

File: bindings/rascal/utils/filter.py [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+            map_by_manager[sp],
+        )
+        for ii, selected_idx in zip(ids, selected_ids):
+            selected_idx.extend(ii)
+    for ii in range(len(selected_ids)):
+        selected_ids[ii] = list(np.sort(selected_ids[ii]))
+    return selected_ids
+
+
+def convert_selected_global_index2perstructure_index(
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:167]: Agreed, we probably want to think of a shorter name (and make this module-private with a leading underscore)

File: bindings/rascal/utils/filter.py [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+
+    Nselect: int
+        number of points to select. If act_on='sample per species' then it should
+        be a dictionary mapping atom type to the number of samples, e.g.
+        Nselect = {1:200,6:100,8:50}.
+
+    act_on: string
+        Select how to apply the selection. Can be either of 'sample',
+        'sample per species','feature'.
+        For the moment only 'sample per species' is implemented.
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:229]: Yes, a lot of this was simply copied from the old `FPSFilter` and `CURFilter` classes, so there's a lot of stuff that escaped earlier reviews. But we can take care of the obvious stuff here too, I suppose.

File: bindings/rascal/utils/filter.py [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+            of selection indices, in the order that they were selected,
+            when `indices=True` and `ordered=True`.
+    """
+
+    def __init__(
+        self,
+        representation,
+        Nselect,
+        act_on="sample per species",
+        selector=None,
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:244]: No idea, I suppose it's so that you can change the selector later... but I don't see a use case for that right now.

File: bindings/rascal/utils/filter.py [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+
+                    in_sample_indices = np.array(
+                        self._selector[sp].get_support(indices=True, ordered=True),
+                        dtype=int,
+                    )
+                    self.selected_sample_ids_by_sp[sp] = in_sample_indices
+                else:
+                    self.selected_sample_ids_by_sp[sp] = []
+
+            return self
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:327]: Right, I think the docstring is wrong. `filter()` returns `SparsePoints` or a list of feature indices, but `select()` just returns itself (is this the sklearn convention?)

File: bindings/rascal/utils/filter.py [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+            self.selected_ids = convert_selected_global_index2perstructure_index(
+                managers, selected_ids_global, strides, map_by_manager
+            )
+            return self.selected_ids
+            # SparsePoints is not compatible with a non center atom species
+            # dependant sparse points
+            # # build the pseudo points
+            # pseudo_points = SparsePoints(self._representation)
+            # pseudo_points.extend(managers, self.selected_ids)
+            # return pseudo_points
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:416]: I think this is the clue to your question a few lines above. One would _like_ this branch to return a SparsePoints, but for some "internal" reason (🤦) you can't make a SparsePoints if you don't do the selection by species (I think...?).

I'll see if there's a simple fix, but if it turns out SparsePoints is too broken to handle this, then we might need to do an overhaul (out of the scope of this PR).

File: bindings/rascal/utils/filter.py [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+                    self.selected_ids[key].append(int(coef_idx[key]))
+            # keep the global indices and ordering for ease of use
+            self.selected_ids[
+                "selected_feature_ids_global"
+            ] = selected_feature_ids.tolist()
+            self.selected_ids[
+                "selected_feature_ids_global_selection_ordering"
+            ] = selected_ids_sorting.tolist()
+            self.selected_ids = dict(coefficient_subselection=self.selected_ids)
+            return self.selected_ids
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:441]: In this case it makes no sense to return a `SparsePoints`, since the selection is done on features. What's returned here is a dictionary you can use to update the hypers and construct a new representation that computes the features sparsely (see the new example notebook for usage). But the docstring should definitely be updated to reflect this.

File: bindings/rascal/utils/filter.py [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+
+        super().__init__(
+            representation=representation,
+            Nselect=Nselect,
+            act_on=act_on,
+            selector=selector,
+            **kwargs,
+        )
+
+    def plot_fps_distances(self):
```
[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:542]: Again, inherited from the old `FPSFilter` interface; I remember a discussion on exactly that point - let me see if I can find it

[the remark this answers is not in the document]
Comment (max-veit) [reply bindings/rascal/utils/filter.py:542]: Here: [#265 (comment)](https://github.com/cosmo-epfl/librascal/pull/265#discussion_r423726259).

But I don't think that's a sufficient justification for pulling in matplotlib as a dependency and basically just wrapping the `plot()` call for the user. I'll take out the plotting bit and just make it return the distances.

File: bindings/rascal/utils/filter.py [hunk 9]
```diff
@@ -504,7 +504,6 @@ def __init__(
             }
 
         else:
-            assert act_on == "feature"
```
Comment (Luthaf) [inline bindings/rascal/utils/filter.py]: I think we want to keep this. Even if the validity check is done in the super class we are still assuming this to be true in that class. This assert would catch case like us adding a new "act_on" mode and forgetting to update this specific class

File: bindings/rascal/utils/filter.py [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+                n: _FPS(
+                    selection_type="sample", n_to_select=Nselect[n], **selector_args
+                )
+                if Nselect[n] > 0
+                else None
+                for n in Nselect
+            }
+
+        else:
+            selector = _FPS(
```
Comment (Luthaf) [inline bindings/rascal/utils/filter.py:544]: Same about having an assert around here to catch future updates to `act_on`

File: bindings/rascal/utils/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply bindings/rascal/utils/__init__.py]: Yes, getting rid of the corresponding C++ code is a good idea!

File: bindings/rascal/utils/filter.py (hunk 9 above)
Comment (max-veit) [reply bindings/rascal/utils/filter.py]: Yeah, but it was breaking the tests, so I took it out. I don't like having two separate checks that each look like copypasta of each other.

Comment (Luthaf) [reply bindings/rascal/utils/filter.py]: I prefer writing assert as defensively as possible, but I can see how you could have different preferences here =). I think we should fix the test instead of changing the code, but I can leave with this as well.

Comment (max-veit) [reply bindings/rascal/utils/filter.py]: I agree with the concept of defensive programming in general; I just don't like to do it with asserts because the error messages are so unhelpful (compare this with the ValueError raised in the superclass that explicitly tells the user what they did wrong). I guess if we really wanted to do this properly, we'd implement a superclass method `check_mode(self, valid_modes)`, given a list of valid modes, and let each subclass use it to throw the error.
